scripted_render_pipeline/postcorrector/post_corrector.py

```diff
@@ -0,0 +1,257 @@
+        """Get median value of given percentile of select images"""
+        # Collect percentile values
+        ps = []
+        # Loop through tiffs
+        for file_path in filepaths:
+            with tifffile.TiffFile(file_path) as tiff:
+                if not tiff.pages:
+                    raise RuntimeError(f"found empty tifffile: {file_path}")
+                # Read tiff and extract lowest resolution page from pyramid
+                image = tiff.pages[-1].asarray()
+                # Compute percentile
+                p1 = np.percentile(image, pct)
+                ps.append(p1)
+        # Compute median
+        med = np.median(ps)
+        return med
+
+    def get_mad(self, filepaths, med, pct=1):
+        """Get median absolute deviation from given percentile of select images
+
+        References
+        ----------
+        [1] https://en.wikipedia.org/wiki/Median_absolute_deviation
+        """
+        # Collect absolute deviations
+        ads = []
+        # Loop through tiffs
+        for file_path in filepaths:
+            # Read tiff and extract lowest resolution page from pyramid
+            tiff = tifffile.TiffFile(file_path.as_posix())
+            image = tiff.pages[-1].asarray()
+            # Compute absolute deviation
+            p1 = np.percentile(image, pct)
+            ad = np.abs(p1 - med)
+            ads.append(ad)
+        # Compute median absolute deviation
+        mad = np.median(ads)
+        return mad
+
+    def has_artefact(self, image, med: float, mad: float, pct=1, a=3):
+        """Determine if image contains an artefact based on intensity percentiles
+
+        image: Input image
+        med: Median `pct`-percentile value of megafield
+        mad: Median absolute deviation from `pct`-percentile across megafield
+        pct: Percentile
+        a: Scaling factor for thresholding the deviation from the median
+            Increasing `a` will allow for larger deviations
+
+        Returns
+        -------
+        corrupted: Bool. 
+            Whether image has been corrupted by an artefact
+        """
+        p1 = np.percentile(image, pct)
+        corrupted = ((p1 < med - a*mad) | (p1 > med + a*mad))
+        return corrupted
+    
+    def post_correct(self, filepaths: list, fps_clean: list):
+        """Reapply post-processing corrections to images
+
+        Parameters
+        ----------
+        
+        filepaths : Filepaths to raw images in one section
+        fps_clean : List of filepaths of artefact-free fields
+        """
+        sum_of_files = 0.0  # set to float to avoid integer overflow
+        
+        # Set target output directory
+        post_correction_dir = filepaths[0].parent / POST_CORRECTIONS_DIR
+        post_correction_dir.mkdir(parents=True, exist_ok=self.clobber)
+        # Copy metadata because render_import requires it
+        shutil.copyfile(filepaths[0].parent / METADATA_FILENAME, post_correction_dir / METADATA_FILENAME)
```

Comment:
the importing gets the metadata from the parent directory for corrected datasets too, it would be simple to add to the importer so we don't need to copy this file to multiple places.

Reply:
Yeah that is also fine, I think the FAST-EM acquisition software automatically duplicates it for the "corrected" stack  but this is not necessary at all.